metabase/metabase
# Create Admin Setting for MB_SESSION_COOKIE_SAMESITE

**Problem:** In interactive embedding, the browser uses cookies to authenticate embedded users against Metabase. If the embedding app and the Metabase instance are not on the same domain, this won't work, as modern browsers won't use cross-domain cookies by default. So, we are creating an option to set a cookie setting called `SameSite` to allow for cross-domain requests with cookies.

**Context:**
- Closes: https://github.com/metabase/metabase-private/issues/182

---
**Desired Behavior**/**Acceptance Criteria**

_Behaviors_
- [ ] Add a "SameSite cookie setting" to allow admins to change the behavior of cross-site requests for use in cross-domain embedding
- [ ] If any authorized origin is different from the instance domain and SameSite cookie setting is not set to 'none', show a note that they may want to set SameSite cookie setting to 'none'
  ![image](https://github.com/metabase/metabase/assets/22608765/6e64f12c-bcfe-405d-a057-51871cb86182)


_Tests_
- [ ] SameSite cookie setting is not shown in non-paid/EE versions
- [ ] SameSite cookie note is shown only when an authorized origin is different from the instance domain and SameSite is not set to 'none'

---
**How to Verify**

1. Run EE of metabase
2. Sign-in as admin
3. Add token with embedding permissions in admin settings > license and billing
4. Go to admin settings > embedding
5. Enable embedding
6. Go to interactive embedding
7. See that there is now an option to configure the SameSite cookie setting

![image](https://github.com/metabase/metabase/assets/22608765/ec0ef916-a2f9-473d-90cd-b398f06ec401)

8. See that the default value is `"lax"`
9. Change the setting to all different possible combinations `"lax" | "strict" | "none"`
  a. Verify that the setting is correctly updated and saved
  b. Verify that new network requests are sent with a parameter in `Set-Cookie` called `SameSite` which is the same as the SameSite cookie setting

![Screenshot from 2023-11-03 18-46-22](https://github.com/metabase/metabase/assets/22608765/2c4b5f71-0230-41fc-a841-488b3a912aa5)


---

✅ **Manually Verified**

https://github.com/metabase/metabase/assets/22608765/57e21c0b-c768-4552-83ab-1610030d4395

- [x] Manually verify that cross-domain interactive embedding works with SameSite setting set to `SameSite=none`
  - Requirements:
    - Metabase App Setup: i.e. `https://metabase-example-app.com`
      - authorized origins set to embedding app domain: `https://embedding-example-app.com`
      - SameSite cookie set to `SameSite=none`
      - SSO setup for embedding login
    - Embedding App Setup: i.e. `https://embedding-example-app.com`
      - setup to embed metabase app with SSO: `https://metabase-example-app.com`
    - https setup - required for use with `SameSite=none` (otherwise browser will reject requests)
